### PR TITLE
Gvisor with CI-test

### PR DESCRIPTION
**Pull request by nathanieltornow**
This is the follow-up PR of #327, with the aim to add functionality for gVisor-sandboxes with a CI-test. Currently the CI-test, which should work very similar to the CRI-test for Firecracker aren't working (probably because of uncleaned resources), and we're aiming to fix this issue in this PR.

In the attached logs, you can see a run of the gvisor CI-test, which typically failes because Minio doesn't start as it should:

```
2021-08-16T09:21:16.5266688Z - Processing resources for Istio core.
2021-08-16T09:21:16.6008347Z ✔ Istio core installed
2021-08-16T09:21:22.7878349Z - Processing resources for Istiod.
2021-08-16T09:26:16.7975346Z - Processing resources for Istiod. Waiting for Deployment/istio-system/istiod
2021-08-16T09:26:16.7983380Z ✘ Istiod encountered an error: failed to wait for resource: resources not ready after 5m0s: timed out waiting for the condition
2021-08-16T09:26:16.8187458Z Deployment/istio-system/istiod
2021-08-16T09:26:17.0547097Z - Processing resources for Addons, Ingress gateways.
2021-08-16T09:31:17.0922398Z - Processing resources for Addons, Ingress gateways. Waiting for Deployment/istio-system/cluster-...
2021-08-16T09:31:17.0926198Z ✘ Ingress gateways encountered an error: failed to wait for resource: resources not ready after 5m0s: timed out waiting for the condition
2021-08-16T09:31:17.0928996Z Deployment/istio-system/cluster-local-gateway
2021-08-16T09:31:17.2184866Z Deployment/istio-system/istio-ingressgateway
2021-08-16T09:31:25.2275580Z - Processing resources for Addons. Waiting for Deployment/istio-system/istiod
2021-08-16T09:31:25.2279316Z ✘ Addons encountered an error: failed to wait for resource: resources not ready after 5m0s: timed out waiting for the condition
2021-08-16T09:31:25.2282316Z Deployment/istio-system/istiod
2021-08-16T09:31:27.9852542Z - Pruning removed resourcesError: failed to install manifests: errors occurred during operation
```

[ctrd-logs.zip](https://github.com/ease-lab/vhive/files/6992317/ctrd-logs.2.zip)

[CI-logs.zip](https://github.com/ease-lab/vhive/files/6992313/CI-logs.zip)

## Reproduce:

- Start a runner
    ```shell
    # Set up the host (the same script as for the self-hosted GitHub CI runner)
    ./scripts/github_runner/setup_runner_host.sh
    # If the $HOSTNAME contains characters which are not allowed in Github runner name, change it to something concise.
    export HOSTNAME=cloudlab
    # start the runner and connect it to github
    # More than 1 runners per machine do not work currently
    /scripts/github_runner/start_runners.sh 1 https://github.com/<OWNER>/<REPO> <Github Access key> cri
    ```
- Trigger the workflow on the `nathanieltornow:gvisor-ci` branch

## Reproduce manually inside a worker:

- Start a runner
    ```shell
    # Set up the host (the same script as for the self-hosted GitHub CI runner)
    ./scripts/github_runner/setup_runner_host.sh
    # If the $HOSTNAME contains characters which are not allowed in Github runner name, change it to something concise.
    export HOSTNAME=cloudlab
    # start the runner and connect it to github
    # More than 1 runners per machine do not work currently
    /scripts/github_runner/start_runners.sh 1 https://github.com/<OWNER>/<REPO> <Github Access key> cri
    ```
- Go into the container `docker ps` -> copy containerID -> `docker exec -it <containerID> bash`
- In the container do:
    ```shell
    cd
    git clone https://github.com/nathanieltornow/vhive.git --branch gvisor-ci
    cd vhive
    ./scripts/install_go.sh
    ./scripts/setup_gvisor_containerd.sh
    mkdir -p /etc/vhive-cri # (to make sure the directory of the cri-socket exists)
    
    # exit and go into the container to make sure to have a fresh session after installation
    exit 
    docker exec -it <containerID> bash
    
    # Run the tests
    sudo -E env "PATH=$PATH" make test-cri-gvisor
    ```
Logs:
[output.txt](https://github.com/ease-lab/vhive/files/7006053/output.txt)
[ctrd-logs-manual.zip](https://github.com/ease-lab/vhive/files/7006058/ctrd-logs-manual.zip)

## The tests are running through on bare metal (see the logs below):

[ctrd-logs-bare_metal.zip](https://github.com/ease-lab/vhive/files/7001059/ctrd-logs-bare_metal.zip)

[bare_metal.out.txt](https://github.com/ease-lab/vhive/files/7001066/bare_metal.out.txt)

### CODE REVIEW

File: scripts/github_runner/clean_cri_runner.sh [hunk 1]
```diff
@@ -36,6 +36,8 @@ sudo pkill -INT vhive
 sudo pkill -9 firecracker-containerd
 sudo pkill -9 firecracker
 sudo pkill -9 containerd
+sudo pkill -9 -f containerd
```
Comment (ustiugov) [inline scripts/github_runner/clean_cri_runner.sh:39]: kind of duplicate of the above

File: vhive.go [hunk 2]
```diff
@@ -44,7 +47,7 @@ const (
 	port    = ":3333"
 	fwdPort = ":3334"
 
-	testImageName = "ghcr.io/ease-lab/helloworld:var_workload"
+	testImageName = "vhiveease/helloworld:var_workload"
```
Comment (ustiugov) [inline vhive.go:50]: please revert

File: vhive.go [hunk 3]
```diff
@@ -80,9 +83,21 @@ func main() {
 	isLazyMode = flag.Bool("lazy", false, "Enable lazy serving mode when UPFs are enabled")
 	criSock = flag.String("criSock", "/etc/firecracker-containerd/fccd-cri.sock", "Socket address for CRI service")
 	hostIface = flag.String("hostIface", "", "Host net-interface for the VMs to bind to for internet access")
-
+	sandbox := flag.String("sandbox", "firecracker", "Determine the sandboxing-technique to use")
```
Comment (ustiugov) [inline vhive.go:86]:
```suggestion
	sandbox := flag.String("sandbox", "firecracker", "Sandbox tech to use, valid options: firecracker, gvisor")
```

**Review by ustiugov — CHANGES_REQUESTED** [review]
minor things to fix. plus check the linter check's errors

File: scripts/github_runner/clean_cri_runner.sh (hunk 1 above)
Comment (nathanieltornow) [reply scripts/github_runner/clean_cri_runner.sh:38]: fixed

File: vhive.go (hunk 2 above)
Comment (nathanieltornow) [reply vhive.go]: fixed

File: vhive.go (hunk 3 above)
Comment (nathanieltornow) [reply vhive.go]: fixed

File: scripts/cloudlab/setup_node.sh [hunk 4]
```diff
@@ -35,6 +35,7 @@ $SCRIPTS/setup_system.sh
 
 if [ "$STOCK_CONTAINERD" != "stock-only" ]; then
     $SCRIPTS/setup_firecracker_containerd.sh
+    $SCRIPTS/setup_gvisor_containerd.sh
```
Comment (shyamjesal) [inline scripts/cloudlab/setup_node.sh:38]: Please add 3 separate cases for stock-containerd, fccd anf gvisor containerd.

File: scripts/cloudlab/start_onenode_vhive_cluster.sh [hunk 5]
```diff
@@ -26,6 +26,17 @@ DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
...
+    SANDBOX="firecracker"
+fi
+
+if [ "$SANDBOX" != "gvisor" ] && [ "$SANDBOX" != "firecracker" ]; then
+    echo Specified sanboxing technique is not supported. Possible are \"firecracker\" and \"gvisor\"
+    exit 1
+fi
+
 echo Clean up host resources if left after previous runs
 $SCRIPTS/github_runner/clean_cri_runner.sh
```
Comment (shyamjesal) [inline scripts/cloudlab/start_onenode_vhive_cluster.sh:41]: Pass gvisor as an argument to clean_cri_runner and add a case for the same [here](https://github.com/nathanieltornow/vhive/blob/d04376524de4c83afd1a54e6149f492efcc97a71/scripts/github_runner/clean_cri_runner.sh#L30).

File: scripts/cloudlab/start_onenode_vhive_cluster.sh [hunk 6]
```diff
@@ -36,16 +47,22 @@ echo Run the stock containerd daemon
...
 echo Build vHive
 cd $ROOT
 source /etc/profile && go build
 
 echo Running vHive with \"${GITHUB_VHIVE_ARGS}\" arguments
-sudo ./vhive ${GITHUB_VHIVE_ARGS} 1>$CTRDLOGDIR/orch.out 2>$CTRDLOGDIR/orch.err &
+sudo ./vhive -sandbox ${SANDBOX} ${GITHUB_VHIVE_ARGS} 1>$CTRDLOGDIR/orch.out 2>$CTRDLOGDIR/orch.err &
 sleep 1s
 
 $SCRIPTS/cluster/create_one_node_cluster.sh
```
Comment (shyamjesal) [inline scripts/cloudlab/start_onenode_vhive_cluster.sh:68]: Pass `gvisor` as an argument to `create_one_node_cluster.sh` and add a case for the same [here](https://github.com/nathanieltornow/vhive/blob/0725f4c140693b101d82a62caef0f21070a1c01d/scripts/cluster/create_one_node_cluster.sh#L33)

**Comment by shyamjesal** [comment]
Please add a case for `gvisor` [here](https://github.com/nathanieltornow/vhive/blob/gvisor-ci/scripts/cluster/setup_worker_kubelet.sh) too. This would be used for setting up multi-node clusters

**Comment by nathanieltornow** [comment]
Hi @shyamjesal, thank you for your comments!
I think I made all the changes you suggested. 
For the cases regarding the socket for the vhive-cri (former `/etc/firecracker-containerd/fccd-cri.sock`, which shouldn't be specific to firecracker imo) in the scripts, I now made a socket which is just for the vhive-cri (`/etc/vhive-cri/vhive-cri.sock`) where the vhive-cri-service will be listening.

File: scripts/cloudlab/start_onenode_vhive_cluster.sh (hunk 6 above)
Comment (ustiugov) [reply scripts/cloudlab/start_onenode_vhive_cluster.sh:68]: I resolve it, I think that further modifications of `create_one_node_cluster.sh` are not required.

**Review by ustiugov — DISMISSED** [review]

File: vhive.go (hunk 2 above)
Comment (ustiugov) [reply vhive.go]: not fixed :) check

**Pull request merged**